core/lib/tests/drop-in-async-context.rs

```diff
@@ -0,0 +1,67 @@
+#[macro_use]
+extern crate rocket;
+
+use rocket::{Build, Config, Rocket};
+use rocket::fairing::AdHoc;
+use rocket::figment::Figment;
+
+struct AsyncDropInAsync;
+
+impl Drop for AsyncDropInAsync {
+    fn drop(&mut self) {
+        // Ensure that managed state is dropped inside of an async context by
+        // ensuring that we do not panic when fetching the current runtime.
+        //
+        // Crates like rocket_sync_db_pools spawn tasks to asynchronously
+        // complete pool shutdown which must be done in an async context or else
+        // the spawn will panic. We want to ensure that does not happen.
+        let _ = rocket::tokio::runtime::Handle::current();
+    }
+}
+
+fn rocket() -> Rocket<Build> {
+    let figment = Figment::from(Config::debug_default())
+        .merge(("address", "tcp:127.0.0.1:0"));
+
+    rocket::custom(figment)
+        .manage(AsyncDropInAsync)
+        .attach(AdHoc::on_liftoff("Shutdown", |rocket| Box::pin(async {
+            rocket.shutdown().notify();
+        })))
+}
+
+mod launch {
+    #[launch]
+    fn launch() -> _ {
+        super::rocket()
+    }
+
+    #[test]
+    fn test_launch() {
+        main();
+    }
+}
+
+mod main {
+    #[rocket::main]
+    async fn main() {
+        super::rocket().launch().await.unwrap();
+    }
+
+    #[test]
+    fn test_main() {
+        main();
+    }
+
+    #[test]
+    fn test_execute() {
+        rocket::execute(async {
+            super::rocket().launch().await.unwrap();
+        });
+    }
+
+    #[test]
+    fn test_execute_directly() {
+        rocket::execute(super::rocket().launch()).unwrap();
+    }
```

Comment:
I added this test, which I _think_ should pass for the same reasons?

Reply:
What's happening here, is Rocket is created and launched, on the async runtime. However, launch returns `Result<Rocket<Ignite>, Error>`. In the success case, Rocket is then returned by `rocket::execute`, and then dropped outside of it (and the async runtime). For 0.6, we should consider changing the interface of `rocket::execute` to `fn execute<P>(f: impl Future<Output = Rocket<P>>) -> Result<(), Error>`, and calling launch internally.

Reply:
This is why the one I wrote explicitly calls unwrap inside an async block.

Reply:
The main use-case is exactly calling it with `rocket.launch()` directly, so it should be able to do that without any footguns or extra ceremony.

Reply:
I'm wondering if there's a better approach altogether. In addition to trying to drop `Rocket` in an async context, why not also specifically drop managed state on shutdown? That is, as part of the shutdown procedure, which is async, drop managed state.

Reply:
I think that might be a valid approach, but the tradeoff would be that we can't support relaunching a Rocket application after it's been shutdown.

Given all this, what is the value to Rocket returning the instance after shutdown? The two values I can see is a) relaunching Rocket after shutdown, and b) inspecting managed state after shutdown. The fact that Rocket is dropped in the error case mean it's likely not super useful to inspect managed state, since that's when you would actually want to inspect it. I'm not sure what the exact use case of relaunching Rocket is, and I'm not convinced it wouldn't be better to recommend simply building a new instance of Rocket.

I think the best compromise would be changing launch to not return the Rocket instance, ever. This ensures Rocket is dropped inside an async context (since it's dropped before the launch future completes). Alternatively, we could create a `Shutdown` phase to return, which would have already dropped managed state, and be safe drop outside an async context.